File: coding-challenges/sources/leetcode/642. Design Search Autocomplete System/642-DesignSearchAutocompleteSystem.swift


fileprivate class TrieNode {
    var children: [Character: TrieNode] = [:]
    var sentences: [String: Int] = [:]
}

fileprivate class Trie {
    var root: TrieNode

    init() {
        root = TrieNode()
    }

    func insert(_ sentence: String, _ times: Int) {
        var node = root
        for c in sentence {
            if node.children[c] == nil {
                node.children[c] = TrieNode()
            }
            node = node.children[c]!
            node.sentences[sentence, default: 0] += times
        }
    }

    func search(_ prefix: String) -> [String] {
        var node = root
        for c in prefix {
            guard let nextNode = node.children[c] else {
                return []
            }
            node = nextNode
        }
        let sortedSentences = node.sentences.sorted {
            $0.value > $1.value || ($0.value == $1.value && $0.key < $1.key)
        }
        return sortedSentences.prefix(3).map { $0.key }
    }
}
fileprivate class AutocompleteSystem {
    var trie: Trie
    var currentPrefix: String

    init(_ sentences: [String], _ times: [Int]) {
        trie = Trie()
        currentPrefix = ""

        for (sentence, time) in zip(sentences, times) {
            trie.insert(sentence, time)
        }
    }

    func input(_ c: Character) -> [String] {
        if c == "#" {
            trie.insert(currentPrefix, 1)
            currentPrefix = ""
            return []
        } else {
            currentPrefix.append(c)
            return trie.search(currentPrefix)
        }
    }
}

import XCTest

import XCTest

class AutocompleteSystemTests: XCTestCase {

    func testAutocompleteSystem() {
        let sentences = ["i love you", "island", "iroman", "i love leetcode"]
        let times = [5, 3, 2, 2]
        let autocomplete = AutocompleteSystem(sentences, times)

        XCTAssertEqual(autocomplete.input("i"), ["i love you", "island", "i love leetcode"])

        XCTAssertEqual(autocomplete.input(" "), ["i love you", "i love leetcode"])

        XCTAssertEqual(autocomplete.input("a"), [])

        XCTAssertEqual(autocomplete.input("#"), [])

    }

}

File: coding-challenges/sources/leetcode/642. Design Search Autocomplete System/642-DesignSearchAutocompleteSystem.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.sentences = {}
    
class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, sentence, times):
        node = self.root
        for c in sentence:
            if c not in node.children:
                node.children[c] = TrieNode()
            node = node.children[c]
            if sentence in node.sentences:
                node.sentences[sentence] += times
            else:
                node.sentences[sentence] = times
    
    def search(self, prefix):
        node = self.root
        for c in prefix:
            if c not in node.children:
                return []
            node = node.children[c]
        top_sentences = sorted(node.sentences.items(), key=lambda x: (-x[1], x[0]))[:3]
        return [sentence for sentence, _ in top_sentences]

class AutocompleteSystem:
    
        def __init__(self, sentences, times):
            self.trie = Trie()
            self.current_prefix = ""

            for sentence, t in zip(sentences, times):
                self.trie.insert(sentence, t)
    
        def input(self, c):
            if c == '#':
                self.trie.insert(self.current_prefix, 1)
                self.current_prefix = ""
                return []
            else:
                self.current_prefix += c
                return self.trie.search(self.current_prefix)

import unittest

class TestAutocompleteSyste(unittest.TestCase):
    def test(self):
        system = AutocompleteSystem(["i love you", "island","ironman", "i love leetcode"], [5,3,2,2])
        self.assertEqual(system.input('i'), ["i love you", "island","i love leetcode"])
        self.assertEqual(system.input(' '), ["i love you","i love leetcode"])
        self.assertEqual(system.input('a'), [])
        self.assertEqual(system.input('#'), [])

if __name__ == "__main__":
    unittest.main()